Library Restructure

stealthanthrax opened the pull request:
Aims to fix #4

stealthanthrax commented:
@hongquan @norbusan , please let me know about your opinions on the new structure

norbusan commented:
Hi @stealthanthrax 
what is the aim of this restructuring? The only thing I see is wrapping everything up into a `Class`, but other than that not much has changed. Are there any further changes planned? What is the advantage of having it as a Class and defining various accessor at the top for backward compatibility?

Thanks

stealthanthrax commented:
@norbusan , this removes the issue of having global variables while still having default values.
I could've passed the variables as default variables and removed global variables, but in the functions in the file, almost every function depends on the global variable, i.e. either reads from it or updates it. So, it made much more sense if we convert these variables as props of a class and also I added a function in the constructor that checks for the `env variable` as well and to incorporate that we would still need to follow global variables in case we follow a `functional programming` approach.

hongquan commented:
Don't name the main class `Main`. It is too general and people looking to an application code will not know where that class comes from.

See https://pythonhosted.org/github2/api/client.html as an example for how to desgin a lib which calls to a service.

hongquan commented:
This is example of a lib, whose service host can be changed: https://redis-py.readthedocs.io/en/latest/

hongquan commented:
Don't use `print`. Use logger instead.

norbusan commented:
Haven't I removed the print already ...?

hongquan commented:
> would still need to follow global variables in case we follow a functional programming approach.

You misunderstood what is "functional programming". If that "global variable" is modified (like in old "susi_python" lib), it is against "functional programming".

hongquan commented on susi_python/main.py:
```diff
@@ -3,221 +3,221 @@
+        try:
+            chat_url = 'http://localhost:4000/susi/chat.json'
+            if (requests.get(chat_url, test_params)):
+                print('connected to local server')
```
@stealthanthrax 

`print` is still there.

hongquan commented on susi_python/main.py:
```diff
@@ -3,221 +3,221 @@
+            device_info_response = requests.get(get_device_info,param1)
+            device_info = device_info_response.json()
+
+        # print(device_info)
```
What is the purpose of this commented line?

hongquan commented on susi_python/main.py:
```diff
@@ -3,221 +3,221 @@
+
+        if device_info is not None:
+            device = device_info['devices'] # list of existing mac ids
+            print(device)
```
Still `print`.

stealthanthrax replied:
@hongquan , I can't see any logger logs integrated in this repo. I'll still check it out.

hongquan commented:
@stealthanthrax About integrating logger, you can just do this:

```py
import logging


logger = logging.getLogger(__name__)
```

The application then will decide how to display this logger.

stealthanthrax commented:
Thanks @hongquan , understood it now. Will do the changes 🖖

stealthanthrax commented:
@hongquan , I have updated the changes, please have a look.

norbusan commented on susi_python/main.py:
```diff
@@ -134,26 +136,26 @@ def _get_rss_entities(self, data):
         }
 
-        # print(access_token)
+        logger.debug(access_token)
```
Not sure how access token are used here, but isn't this creating the danger that a users shared the debug log and thus publicizes the access token?

stealthanthrax replied:
@norbusan , it was being used when we trying to ad offline/online support. I don't think we need it anymore. So , I should probably remove it.

norbusan replied:
Yes please

norbusan replied:
I think this change is still open @stealthanthrax

norbusan commented on susi_python/main.py:
```diff
@@ -5,11 +5,13 @@
+logger = logging.getLogger(__name__)
 
-class Main:
+class SusiMain:
```
Sorry to be so picky, but again I think `SusiMain` is not a good naming. This module does nothing **main**, just provides access to the SUSI Server. I think the proper name would be `SusiServer`!

Why, look at the code example. Currenty we would have
```
import susi_python
susi_main.ask('...')
```
which sounds strange, because the **main** part to me is `susi_linux`. If we use `SusiServer` and `susi_server` instead, the code would look like:
```
import susi_python
susi_server.ask('....')
```
which is somehow *what we are doing* ...

WDYT?

stealthanthrax replied:
@norbusan , susi_server is refered to the Java Server here. So, might cause a lot of confusions here.
Should I name it just `SUSI`?

norbusan replied:
But it is about connecting to the susi server, right. It is so to say a local tunnel to the server.

But ok, SUSI/susi alone is fine with me!

hongquan replied:
I think we can name it `SUSIClient`

stealthanthrax replied:
Yes, `SUSIClient` sounds good to me.

norbusan replied:
Ok, fine with me, whatever. Please go ahead and commit the changes.

stealthanthrax replied:
@norbusan @hongquan , updated

norbusan commented:
Ok, looks fine to me. 

Now for a last thing: Why do we need the global variables? I `git grep susi_python` in the `susi_linux` sources and it is only used in two instances, where we can adjust it to use `susi.susi_client...`

That would make everything even more clean ... wdyt @hongquan @stealthanthrax  ?

hongquan commented:
@norbusan I don't understand what you mean.
Aren't we trying to drop global variables?

stealthanthrax commented:
@norbusan , I don't see any global variables anymore. Can you direct us to the file that you are talking about?

norbusan commented:
@hongquan @stealthanthrax what I mean is the addtion
```
use_api_endpoint = susi_client.use_base_url
query = susi_client.query
ask = susi_client.ask
answer_from_json = susi_client.answer_from_json
sign_in = susi_client.sign_in
sign_up = susi_client.sign_up
forgot_password = susi_client.forgot_password
get_previous_responses = susi_client.get_previous_responses
update_location = susi_client.update_location
```
I though it would be nice to get rid of these, too ;-)

hongquan commented:
@norbusan I agree to get rid of them.

stealthanthrax commented:
Yes, that makes sense. I'll get rid of them.

norbusan commented on susi_python/__init__.py:
```diff
@@ -1,11 +1 @@
-sign_up = susi_client.sign_up
-forgot_password = susi_client.forgot_password
-get_previous_responses = susi_client.get_previous_responses
-update_location = susi_client.update_location
```
Removing these variables isn't enough, because it will break `susi_linux` and maybe other clients. We need to search for uses of this library within the susi environment, and provide patches that remove the use of these vars. Only **after** the uses have been fixed, we can actually remove that!

stealthanthrax replied:
okay @norbusan , lets just keep it as wip then and I'll open a dependant issue in SUSI Linux and then we'll merge them simultaneously?

norbusan replied:
Yes!

stealthanthrax commented:
@norbusan , I have updated and tested it on a pc

norbusan commented:
@stealthanthrax I don't know what you have tested, but this is broken. Completely.

Look at the answers provided:

Before merge:
```
$ python3
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import susi_python as susi
connected to local server
>>> susi.ask("play some music")
{'identifier': 'ytd-Y-FhDScM_2w', 'language': 'de', 'answer': 'Playing 소유(SoYou) X 정기고(JunggiGo) - 썸(Some) feat. 긱스 릴보이 (Lil Boi of Geeks) M/V'}
>>> 
$
```

After merge
```
$  python3
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from susi_python import susi_client as susi
>>> susi.ask("play some music")
{'identifier': 'ytd-Y-FhDScM_2w'}
>>> 
```

Stuff has completely changed. The answers are not properly dealt with. I am not sure **what** you have tested, but the moment I tried these two (susi_linux and susi_python PRs after merging) it glared at me that it isn't working as before.

I will revert both changes. Please check again. Thanks
